lib/internal/util/inspect.js

```diff
@@ -593,10 +597,26 @@ function isInstanceof(object, proto) {
   }
 }
 
+// Special-case for some builtin prototypes in case their `constructor` property has been tampered.
+const wellKnownPrototypes = new SafeMap();
+wellKnownPrototypes.set(ObjectPrototype, { name: 'Object', constructor: Object });
+wellKnownPrototypes.set(FunctionPrototype, { name: 'Function', constructor: Function });
+
 function getConstructorName(obj, ctx, recurseTimes, protoProps) {
   let firstProto;
   const tmp = obj;
   while (obj || isUndetectableObject(obj)) {
+    const wellKnownPrototypeNameAndConstructor = wellKnownPrototypes.get(obj);
+    if (wellKnownPrototypeNameAndConstructor != null) {
+      const { name, constructor } = wellKnownPrototypeNameAndConstructor;
+      if (FunctionPrototypeSymbolHasInstance(constructor, tmp)) {
+        if (protoProps !== undefined && firstProto !== obj) {
+          addPrototypeProperties(
+            ctx, tmp, firstProto || tmp, recurseTimes, protoProps);
+        }
+        return name;
+      }
+    }
```

Comment:
I fear this might have a performance impact. Could we benchmark the change to see how it influences average cases?
Is the special casing for only `Object` and `Function` also sufficient? We could theoretically track all built-ins, if we want to do this. But what happens if the prototype is no built-in?

Reply:
We have to draw the line somewhere, I don't think we can avoid special casing `Object` (otherwise we wrongfully report as null-prototype all objects as soon as the `%Object.prototype%.constructor` property has been replaced with a getter); special casing `Function` is arbitrary, we have to draw the line somewhere, and I don't feel strongly about it. Let's see what the benchmark says

Reply:
Given that the benchmark CI shows this PR performs better than `main` overall, I suggest we keep `Object` and `Function` for now, and if there's a need for including more built-in prototypes to the list, we can discuss that in follow-up PRs.

Reply:
I guess the performance would also improve further for other objects with more entries. The overall benefit is definitely there for the average case, so that's a great improvement!

I also agree that we have to draw the line somewhere and I am fine with the approach.